test/e2e/network/dns_configmap.go

```diff
@@ -219,6 +228,63 @@ func (t *dnsPtrFwdTest) run() {
 		moreForeverTestTimeout)
 }
 
+type dnsExternalNameTest struct {
+	dnsTestCommon
+}
+
+func (t *dnsExternalNameTest) run() {
+	t.init()
+
+	t.createUtilPod()
+	defer t.deleteUtilPod()
+
+	fooHostname := "foo.example.com"
+	t.createDNSServer(map[string]string{
+		fooHostname: "192.0.2.123",
+	})
+	defer t.deleteDNSServerPod()
+
+	f := t.f
+	serviceName := "dns-externalname-upstream-test"
+	externalNameService := framework.CreateServiceSpec(serviceName, googleDnsHostname, false, nil)
+	if _, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(externalNameService); err != nil {
+		Fail(fmt.Sprintf("Failed when creating service: %v", err))
+	}
+	serviceNameLocal := "dns-externalname-upstream-local"
+	externalNameServiceLocal := framework.CreateServiceSpec(serviceNameLocal, fooHostname, false, nil)
+	if _, err := f.ClientSet.CoreV1().Services(f.Namespace.Name).Create(externalNameServiceLocal); err != nil {
+		Fail(fmt.Sprintf("Failed when creating service: %v", err))
+	}
+	defer func() {
+		By("deleting the test externalName service")
+		defer GinkgoRecover()
+		f.ClientSet.CoreV1().Services(f.Namespace.Name).Delete(externalNameService.Name, nil)
+		f.ClientSet.CoreV1().Services(f.Namespace.Name).Delete(externalNameServiceLocal.Name, nil)
+	}()
+
+	t.checkDNSRecordFrom(
+		fmt.Sprintf("%s.%s.svc.cluster.local", serviceName, f.Namespace.Name),
+		func(actual []string) bool {
+			return len(actual) >= 1 && actual[0] == googleDnsHostname+"."
+		},
+		"dnsmasq",
+		moreForeverTestTimeout)
+
+	t.setConfigMap(&v1.ConfigMap{Data: map[string]string{
+		"upstreamNameservers": fmt.Sprintf(`["%v"]`, t.dnsServerPod.Status.PodIP),
+	}})
+
+	t.checkDNSRecordFrom(
+		fmt.Sprintf("%s.%s.svc.cluster.local", serviceNameLocal, f.Namespace.Name),
+		func(actual []string) bool {
+			return len(actual) == 2 && actual[0] == fooHostname+"." && actual[1] == "192.0.2.123"
+		},
+		"dnsmasq",
+		moreForeverTestTimeout)
+
+	t.c.CoreV1().ConfigMaps(t.ns).Delete(t.name, nil)
```

Comment:
Instead of deleting the configmap, what about simply resetting it to empty? Otherwise we will waste ~1 minute for addon-manager to re-assert it.

Reply:
Good one. Fixed.